sql/distsql: stream decoding and related changes

RaduBerinde opened the pull request:
See individual commits for descriptions.
@andreimatei @paperstreet @nvanbenschoten

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6844)

<!-- Reviewable:end -->

tamird commented on sql/distsql/server_test.go:
```diff
@@ -84,7 +85,20 @@ func TestServer(t *testing.T) {
+	}
+	if done, err := decoder.IsDone(); !done {
+		t.Fatalf("stream not done")
+	} else if err != nil {
```
this case should come first - seems likely that it is unreachable currently.

RaduBerinde commented:
Review status: 0 of 13 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[sql/distsql/server_test.go, line 96 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-KIJAV1zWWr48zKWlaLz-r5-96:-KIOVK2deif7Ez0qYWJ7:820730289) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f82fda5cca9d7f57d57a6e3f2462d026997b9ab/sql/distsql/server_test.go#L96)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this case should come first - seems likely that it is unreachable currently.
</details>
the error return value only makes sense when done is true. I will rename the `err` to `trailerErr` since it's not the "usual" kind of error. The semantic is also not very clear from the IsDone doc, I will fix.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Latest commits in r6 through r10 (r1 through r5 can be ignored).

<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> #### sql/distsql: stream decoding and related changes
> 
> See individual commits for descriptions.
> 
> @andreimatei @paperstreet @nvanbenschoten
> 
> 
> 
</details>


---

Review status: 0 of 13 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-:-KIOnZA6-34KV64-Ni7p:-468749968)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Morning bump.

paperstreet commented:
LGTM. I really like how all this is shaping up. Top notch work

<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Morning bump.
</details>


---

Review status: 0 of 13 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[sql/distsql/outbox.go, line 91 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-KIYFHhD5E1I04qJTQHM:-KIYFHhD5E1I04qJTQHN:-1768851778) ([raw file](https://github.com/cockroachdb/cockroach/blob/620b6b31a47a3d9cbdcfe7f11eb35b7495639e9e/sql/distsql/outbox.go#L91)):*
> ```Go
> }
> 
> // AddRow encodes a row into rowBuf. If enough rows were accumulated
> ```

addRow

---

*[sql/distsql/stream_decoder.go, line 24 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-KIYHjy21lLkcLuVy_IZ:-KIYHA_-Q40kIHN8_3wt:-1790246065) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8ddfaed3caafa6a24f2f8f23228edb945ba84fd/sql/distsql/stream_decoder.go#L24)):*
> ```Go
> )
> 
> // StreamDecoder converts a sequence of StreamMessage to EncDatum rows.
> ```

It'd be nice if this and StreamEncoder had a rowFetcher style example of how to use the methods.

---

*[sql/distsql/stream_decoder.go, line 80 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-KIYI860iIVRl5zCULaa:-KIYI860iIVRl5zCULab:-69719186) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8ddfaed3caafa6a24f2f8f23228edb945ba84fd/sql/distsql/stream_decoder.go#L80)):*
> ```Go
> 	}
> 	for i := range rowBuf {
> 		var err error
> ```

nit: Pull err out of the loop

---

*[sql/distsql/stream_encoder.go, line 34 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-KIYGHRw0TXK2CRoPaC0:-KIYGHRw0TXK2CRoPaC1:-2087631424) ([raw file](https://github.com/cockroachdb/cockroach/blob/620b6b31a47a3d9cbdcfe7f11eb35b7495639e9e/sql/distsql/stream_encoder.go#L34)):*
> ```Go
> 	firstMessageDone bool
> 	alloc            sqlbase.DatumAlloc
> 	// Placeholders to avoid allocations
> ```

nit: They're not really placeholders

---

*[sql/distsql/stream_encoder.go, line 72 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-KIYG2smQ59fApB1e9Cr:-KIYG2snuo-2LkKS1ARE:1999990979) ([raw file](https://github.com/cockroachdb/cockroach/blob/620b6b31a47a3d9cbdcfe7f11eb35b7495639e9e/sql/distsql/stream_encoder.go#L72)):*
> ```Go
> // error.
> //
> // If the message is used (sent), NextMessage must be called. The returned
> ```

"NextMessage must be called before AddRow is used again." or something similiar? Doesn't seem like you have to call NextMessage if you're done calling AddRow

---

*[sql/distsql/tablereader_test.go, line 45 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-KIYD1qs9DUjTY0r7bAm:-KIYD1qtYD3aMi8Yvrjm:379857483) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba9f45571043216f48590f6a25e67c7fd7b9affb/sql/distsql/tablereader_test.go#L45)):*
> ```Go
> 
> func (tr *testingReceiver) PushRow(row sqlbase.EncDatumRow) bool {
> 	rowCopy := append(sqlbase.EncDatumRow(nil), row...)
> ```

Interesting. How does this compare to `rowCopy := make(sqlbase.EncDatumRow, len(row); copy(rowCopy, row)` ? Is this more idiomatic and I've been doing it wrong?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-:-KIYJMxsuGKcv4K0L4Zy:-1120228724)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql/distsql: adding type to DatumInfo

I didn't include the type in DatumInfo because I thought it would be hard to
populate based on the first row (because of potential DNulls). But since we are
using EncDatums which contain the type that is actually not a problem. This will
make it possible to have generic code that decodes a stream.

RaduBerinde committed:
sql/distsql: cleanup around EncDatum and rows of EncDatum

 - moving row type to sqlbase.EncDatumRow.
 - adding EncDatumRows type.
 - adding String methods to EncDatum, EncDatumRow, EncDatumRows
 - simplifying tablereader_test to use EncDatumRows directly
 - exporting RandColumnType and RandDatumEncoding (for distsql tests)

RaduBerinde committed:
sql/distsql: stream decoder

StreamDecoder interprets StreamMessages and gives us rows of EncDatums.

Writing test for StreamEncoder/Decoder.

RaduBerinde committed:
sql/distsql: separate out stream encoding code

We will have a stream decoder and having a standalone stream encoder will be
useful for testing.

RaduBerinde committed:
sql/distsql: server test now verifies output stream

RaduBerinde commented:
TFTR!  Updated - current commits are r11-r15

<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> LGTM. I really like how all this is shaping up. Top notch work
</details>


---

Review status: 0 of 13 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[sql/distsql/stream_decoder.go, line 24 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-KIYHjy21lLkcLuVy_IZ:-KIYMCZsiM8XvsN_WJUw:-1504763431) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8ddfaed3caafa6a24f2f8f23228edb945ba84fd/sql/distsql/stream_decoder.go#L24)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> It'd be nice if this and StreamEncoder had a rowFetcher style example of how to use the methods.
</details>
Good idea.

---

*[sql/distsql/stream_decoder.go, line 80 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-KIYI860iIVRl5zCULaa:-KIYMFQ3pP3pFRHTH6II:903971980) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8ddfaed3caafa6a24f2f8f23228edb945ba84fd/sql/distsql/stream_decoder.go#L80)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> nit: Pull err out of the loop
</details>
why? we should minimize the scope of variables

---

*[sql/distsql/stream_encoder.go, line 72 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-KIYG2smQ59fApB1e9Cr:-KIYMtwMYxuAwtJXamf1:8006014) ([raw file](https://github.com/cockroachdb/cockroach/blob/620b6b31a47a3d9cbdcfe7f11eb35b7495639e9e/sql/distsql/stream_encoder.go#L72)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> "NextMessage must be called before AddRow is used again." or something similiar? Doesn't seem like you have to call NextMessage if you're done calling AddRow
</details>
I see your point. But it's also ok to call `FormMessage` again without calling `NextMessage` if you didn't use the message from the last call (e.g. you couldn't send it). Updated, hopefully it's clearer now.

---

*[sql/distsql/tablereader_test.go, line 45 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-KIYD1qs9DUjTY0r7bAm:-KIYONn1sATjFG2R1cZw:-1900882660) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba9f45571043216f48590f6a25e67c7fd7b9affb/sql/distsql/tablereader_test.go#L45)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> Interesting. How does this compare to `rowCopy := make(sqlbase.EncDatumRow, len(row); copy(rowCopy, row)` ? Is this more idiomatic and I've been doing it wrong?
</details>
I think it's the same thing, this is just shorter. We do it quite a bit. Sometimes it happens implicitly when we append to an uninitialized slice.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-:-KIYPJMR5lvBYHTzvsl_:-839667360)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Reviewed 7 of 7 files at r1, 3 of 3 files at r2, 2 of 2 files at r3, 2 of 2 files at r4, 1 of 1 files at r5, 2 of 3 files at r12, 2 of 2 files at r13, 2 of 2 files at r14, 1 of 1 files at r15.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[sql/distsql/stream_decoder.go, line 38 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-KIYQ3mgPIRf0c57rhP7:-KIYQ3mgPIRf0c57rhP8:1705878209) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e43c5d3820b4c6c80711824fb70804228d4b8c9/sql/distsql/stream_decoder.go#L38)):*
> ```Go
> 	}
> 	if msg.Header != nil {
> 		if sd.info != nil {
> ```

pulling `headerReceived := sd.info != nil` up top would make some of this logic more clear.

---

*[sql/distsql/stream_decoder.go, line 54 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-KIYQWHcQLMTKIN0YHx1:-KIYQWHcQLMTKIN0YHx2:-1421002077) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e43c5d3820b4c6c80711824fb70804228d4b8c9/sql/distsql/stream_decoder.go#L54)):*
> ```Go
> 	if len(msg.Data.RawBytes) > 0 {
> 		if len(sd.data) == 0 {
> 			sd.data = msg.Data.RawBytes
> ```

If this is safe to do, we should probably at least document it.

---

*[sql/distsql/stream_encoder.go, line 69 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-KIYNXF9qEPS6uUyIB_A:-KIYNXFAWJIJyTM8zOC5:1769549863) ([raw file](https://github.com/cockroachdb/cockroach/blob/8557f00b31067eba83c5addc48567a46ae4dc78e/sql/distsql/stream_encoder.go#L69)):*
> ```Go
> 
> // FormMessage populates a message. The StreamMessage should be treated as
> // immutable.  If final is true, a message trailer is populated with the given
> ```

Extra space after `immutable.`

---

*[sql/distsql/stream_encoder.go, line 97 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-KIYN4wG7thaQxVc1Fc4:-KIYN4wH1kUEkvQCaxty:-701190968) ([raw file](https://github.com/cockroachdb/cockroach/blob/8557f00b31067eba83c5addc48567a46ae4dc78e/sql/distsql/stream_encoder.go#L97)):*
> ```Go
> 
> // NextMessage resets the row data for the next message.
> func (se *StreamEncoder) NextMessage() {
> ```

I'm not sure this name describes the method very well. Maybe `ResetForNextMessage`, or something shorter that still gets the point across that this is not returning the next message.

---

*[sql/sqlbase/encoded_datum.go, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-KIYKl11cSzhaXB1Bclw:-KIYKl12sXAtfisrOrKw:1818152838) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea556ed5d861fb769fa9da19d693950f2c2d468d/sql/sqlbase/encoded_datum.go#L48)):*
> ```Go
> 	if ed.Datum == nil {
> 		if a == nil {
> 			a = &DatumAlloc{}
> ```

Interesting, does `EncDatum.Decode` not handle nil `DatumAlloc`? I'm not positive about how this method is used, but it seems like it should either not allocate a `DatumAlloc` if none is provided, or it should return the one it creates so that it can be reused if needed (see `EncDatumRow.stringToBuf` if not given an alloc). 

Creating an alloc internally for every call here seems like an anti-pattern, so it should probably be discouraged. You understand the usage better though, so I'll defer to your judgement.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844)*
<!-- Sent from Reviewable.io -->

paperstreet commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[sql/distsql/stream_decoder.go, line 80 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-KIYI860iIVRl5zCULaa:-KIYXOMSSzi_LKWf4EjK:455495335) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8ddfaed3caafa6a24f2f8f23228edb945ba84fd/sql/distsql/stream_decoder.go#L80)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> why? we should minimize the scope of variables
</details>
I was making an incorrect assumption about perf. Nevermind.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Current commits r16-r20.

<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> TFTR!  Updated - current commits are r11-r15
</details>


---

Review status: 6 of 13 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[sql/distsql/stream_encoder.go, line 97 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-KIYN4wG7thaQxVc1Fc4:-KIYfOPuqwvCx-p-dIcV:-464932530) ([raw file](https://github.com/cockroachdb/cockroach/blob/8557f00b31067eba83c5addc48567a46ae4dc78e/sql/distsql/stream_encoder.go#L97)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> I'm not sure this name describes the method very well. Maybe `ResetForNextMessage`, or something shorter that still gets the point across that this is not returning the next message.
</details>
Looks like this API is causing some confusion, and we don't really use this capability of "not using" a message, so I removed this and moved the code in `FormMessage`.

---

*[sql/sqlbase/encoded_datum.go, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-KIYKl11cSzhaXB1Bclw:-KIYfiHcFC7OGaZr1BYH:38332733) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea556ed5d861fb769fa9da19d693950f2c2d468d/sql/sqlbase/encoded_datum.go#L48)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Interesting, does `EncDatum.Decode` not handle nil `DatumAlloc`? I'm not positive about how this method is used, but it seems like it should either not allocate a `DatumAlloc` if none is provided, or it should return the one it creates so that it can be reused if needed (see `EncDatumRow.stringToBuf` if not given an alloc). 
> 
> Creating an alloc internally for every call here seems like an anti-pattern, so it should probably be discouraged. You understand the usage better though, so I'll defer to your judgement.
</details>
`Decode` needs a `DatumAlloc` to pass to `DecodeTableKey`. I think it's better if it requires one to avoid accidentally passing a nil alloc in a hot path.

This is only used for `String()` and it shouldn't be performance-sensitive, it added it for testing/debugging. I chose to do the allocation here instead of just calling `ed.stringWithAlloc(&DatumAlloc{})` to avoid the allocation if the datum is already decoded.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-:-KIYhZfJSpuqH1t2vzdr:1717565796)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 7 of 7 files at r1, 3 of 3 files at r2, 2 of 2 files at r3, 2 of 2 files at r4, 1 of 1 files at r5, 7 of 7 files at r6, 3 of 3 files at r7, 2 of 2 files at r8, 2 of 2 files at r9, 1 of 1 files at r10, 7 of 7 files at r11, 3 of 3 files at r12, 2 of 2 files at r13, 2 of 2 files at r14, 1 of 1 files at r15, 7 of 7 files at r16, 3 of 3 files at r17, 2 of 2 files at r18, 2 of 2 files at r19, 1 of 1 files at r20.
Review status: all files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[sql/sqlbase/util.go, line 85 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-KIZ-w_7viPoM8DLcCJJ:-KIZ-w_7viPoM8DLcCJK:-1150858945) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea556ed5d861fb769fa9da19d693950f2c2d468d/sql/sqlbase/util.go#L85)):*
> ```Go
> 
> // RandEncDatum generates a random EncDatum (of a random type).
> func RandEncDatum(rng *rand.Rand) EncDatum {
> ```

move these functions to sqlbase/testutils.go ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Current commits r16-r20.
</details>


---

Review status: all files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[sql/distsql/stream_decoder.go, line 49 \[r19\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-KIZ4ODd-9AVJDuCcLB7:-KIZ4ODd-9AVJDuCcLB8:1235989858) ([raw file](https://github.com/cockroachdb/cockroach/blob/834f2606905a6f02b7d65919b555990a7ce965cd/sql/distsql/stream_decoder.go#L49)):*
> ```Go
> 	data            []byte
> 	trailerReceived bool
> 	trailerErr      error
> ```

mind adding a comment to `trailerErr`?

---

*[sql/distsql/stream_encoder.go, line 38 \[r18\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-KIZ2rEA9tTa64U0U4OA:-KIZ2rEBjiwwG_mPktgc:-1683963651) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f96e045c6557c73995e3fa0dcdc0908ff0e4102/sql/distsql/stream_encoder.go#L38)):*
> ```Go
> //       // Send out message.
> //       ...
> //       se.NextMesage()
> ```

there's no longer a `NextMessage`? Probably for the better :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-:-KIZ4o58DbZLOF0bzCEv:1417088721)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: move rand datum functions to testutils.go

RaduBerinde commented:
Review status: 5 of 14 files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[sql/distsql/stream_decoder.go, line 49 \[r19\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-KIZ4ODd-9AVJDuCcLB7:-KIZWGhJVKYoLF9FaAFR:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/834f2606905a6f02b7d65919b555990a7ce965cd/sql/distsql/stream_decoder.go#L49)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> mind adding a comment to `trailerErr`?
</details>
Done.

---

*[sql/distsql/stream_encoder.go, line 38 \[r18\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-KIZ2rEA9tTa64U0U4OA:-KIZWITwB_3U5CeokyMA:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f96e045c6557c73995e3fa0dcdc0908ff0e4102/sql/distsql/stream_encoder.go#L38)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> there's no longer a `NextMessage`? Probably for the better :)
</details>
Done.

---

*[sql/sqlbase/util.go, line 85 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844#-KIZ-w_7viPoM8DLcCJJ:-KIZWQACBKQKqDheVnHj:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea556ed5d861fb769fa9da19d693950f2c2d468d/sql/sqlbase/util.go#L85)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> move these functions to sqlbase/testutils.go ?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6844)*
<!-- Sent from Reviewable.io -->
